### ISSUE TITLE
Top margin incorrect

### ISSUE THREAD

**Issue by drmrbrewer** [highcharts/highcharts#12206]
Take the following example:

https://jsfiddle.net/8wvkstxh/1/

The chart options include a `y` adjustment to the top `xAxis` labels, because, without it, the labels are too far from the axis itself (for my liking):

https://jsfiddle.net/8wvkstxh/2/

But with the `y` adjustment included (first example above), there is now a lot of dead space above the top x axis labels (I'm trying to create a chart layout which minimises dead margin space and maximises chart area, so I'm sensitive to this).

The top margin is seemingly "correct" for the default y position (without the y adjustment -- second example above), in that there is only a small gap between the labels and the edge.

But with the adjusted y position there is too much margin space.

It seems to me that the problem is that in order to set the y position to my liking (to move the labels **down** from their default position), I need a **negative** value for y.  

But a negative y value actually should normally be pushing the labels **up**... so it seems that highcharts is thinking that more margin space will be required, and adds margin space to compensate for this, even though the labels are actually **lower** than without the negative adjustment (so that the **opposite** should have happened... the top margin should have been reduced, not increased).

Compare this with setting y explicitly to `0`:

https://jsfiddle.net/sw36dqtp/

Therein lies the problem: setting `y: 0` gives a very different result to not setting `y` at all.  I need to set `y` to get the y position right, so I need to state a value, and I'm moving it **up** from `y: 0` so I need to specify a negative value, which highcharts is apparently compensating for by adding more margin space which isn't needed.  This leaves a lot of bloat in the top margin.

**Comment by pawelfus** [comment]
Thanks for reporting the issue @drmrbrewer 

In this particular design, you can set [`xAxis.labels.reserveSpace`](https://api.highcharts.com/highcharts/xAxis.labels.reserveSpace) to false, take a look: https://jsfiddle.net/BlackLabel/rw1eag2h/


*Internal note:*
It looks like a bug in padding correction: 
https://github.com/highcharts/highcharts/blob/e62d54124bbe96ce24f8c3357a9d47030d554382/ts/parts/Axis.ts#L6651-L6671

In short: http://jsfiddle.net/BlackLabel/0ou3rfa4/ - both (top/bottom) spacings are set to zero, yet only bottom labels are close enough to the border.

**Comment by drmrbrewer** [comment]
Thanks, @pawelfus.  Is the `reserveSpace` trick an exact fix in my particular example, or does it just reduce the impact?  Because my feeling is that I still want to be "reserving space" for the x axis labels at the top, just not quite so much as is the case at present!  So setting `reserveSpace` to `false` seems wrong, even if it helps.

**Comment by pawelfus** [comment]
It's a direct fix @drmrbrewer - instead of reserving space for labels automatically, you can control it.  Note that labels nicely fit in the space created by ticks. You can reduce extra space above by setting `spacingTop` (defaults to 10px) and `spacingBottom` (defaults to 15px), demo: https://jsfiddle.net/BlackLabel/m9uyvszf/

Extra space, in your demos with titles, is caused by `title.margin` (defaults to 15px). You can disable it, see: https://jsfiddle.net/BlackLabel/f93bn8ck/ - no need to play around with negative adjustments and line-height.

**Comment by drmrbrewer** [comment]
> labels nicely fit in the space created by tickmarks

@pawelfus aaah, I see.  Yes, makes sense!  Thanks.

> Extra space, in your demos with titles, is caused by title.margin

ah yes, thanks... I really hated the negative margin kludge so I'm glad it's not necessary!

> no need to play around with negative adjustments and line-height

but in my case I still need to reduce `line-height` because otherwise I find that the gap between title lines is too big, compare:

https://jsfiddle.net/p8z2476t/
https://jsfiddle.net/p8z2476t/1/

But this is really just particular to my use case where I'm trying to maximise actual chart size as far as possible.

**Comment by pawelfus** [comment]
> I still need to reduce line-height because otherwise I find that the gap between title lines is too big, compare:

Ah, in that case, yes, `line-height` is a good idea :)

**Comment by drmrbrewer** [comment]
@pawelfus this issue was closed automatically but I don't see it as being completely resolved.  I've been using the suggested solution, which has been working fine in general.  Basically the solution was to set `reserveSpace` for the top xAxis labels to `false`, which seems to work only because some space is anyway being reserved for the ticks.

But now I'm in a situation where I want to make the format for the top axis labels customisable, so that I can no longer be sure how much space I need to reserve.  Therefore, using the ticks to reserve space no longer works, but I also can't get the right amount of space to be reserved for the labels... or at least I can't adjust the y position of the labels without messing everything up.

Example: https://jsfiddle.net/vpyakzx3/  (press Run multiple times to show the effect of different label height)

I have deliberately reduced `tickLength` so that it is insufficient to reserve enough space in either scenario.

**Comment by pawelfus** [comment]
Hey @drmrbrewer 
Thank you for the update. In this case (customizable format), I agree you can't use `reserveSpace = false` and rely on ticks' created space. I have removed the code responsible for extra space, awaiting visual tests results ([draft](https://github.com/highcharts/highcharts/pull/17173)).


You can test this branch: `<script src="https://github.highcharts.com/7ba376b/highcharts.js"></script>`


Your demo without workarounds (negative `title.offset` and `labels.y`): https://jsfiddle.net/BlackLabel/7krvqmhx/

**Comment by drmrbrewer** [comment]
Thanks, @pawelfus.  The demo seems to suffer from a similar problem to my original post, in that it seems to be impossible to adjust the `y` position of the labels, because setting an explicit value for `y` messes up the spacing.  

I'm trying to reduce the gap shown here:

![image](https://user-images.githubusercontent.com/9009792/162728095-20942126-50ed-46b0-a69e-8e82c42eb541.png)

So I try to add a small positive `y` adjustment to your demo (the only change), and it results in the following:

https://jsfiddle.net/btx5qhwj/

I'm used to being able to adjust the `x` position of the labels on the yAxis and it seems to work, but here it doesn't.

**Comment by pawelfus** [comment]
Right, back to square one... I think I got it: https://jsfiddle.net/BlackLabel/o5wcehyv/1/ 
Now `y: 0` and `y: undefined` are the same, so any positive `y`-offset will push to the bottom and negative `y`-offset will push to the top. In the demo, I added `title.margin = 0` to reduce the space between the title and xAxis labels.

**Comment by drmrbrewer** [comment]
@pawelfus there still seem to be some quirks.

You say that "Now `y: 0` and `y: undefined` are the same"... this seems to be so for the top (opposite) xAxis.

But now I'm looking at the behaviour of the labels on the bottom xAxis.

Check the following example, which uses the same label setup for top and bottom:

https://jsfiddle.net/bqcj163u/

Hit 'Run' several times... this now changes the `y` for the labels between `0` and `undefined`, and this does change the layout significantly at the bottom (see gap between labels and chart edge):

![image](https://user-images.githubusercontent.com/9009792/163010825-529fe851-604c-494f-b481-f0a706366943.png)

vs:

![image](https://user-images.githubusercontent.com/9009792/163010936-1d295edc-9f6c-4db1-a89b-55521d8100e9.png)

This (i.e. whether `y` of bottom labels is `0` or `undefined`) also seems to have a small effect on the layout at the top (see highlighted gap below, between labels and title):

![image](https://user-images.githubusercontent.com/9009792/163011626-77b53b78-023e-418c-8866-a7d703e79002.png)

vs:

![image](https://user-images.githubusercontent.com/9009792/163011230-8875c4b3-e85e-4764-82b9-8ff3b3bf093d.png)

**Comment by drmrbrewer** [comment]
@pawelfus any further thoughts on this?

**Comment by pawelfus** [comment]
Hi @drmrbrewer 

My apologies for the silence from my side. I will try a different approach next week. It's a quite tricky case, because I don't want to cause any regressions (current PR has plenty of them..)

**Comment by drmrbrewer** [comment]
Hi @pawelfus... no problem, and I completely understand that it is tricky for you, potentially with lots of unintended consequences for any change you might make.  As I mentioned, adjusting the `x` position of the labels on the `yAxis` (or `yAxes`) seems to be very well behaved (somehow magically reserving the right amount of space regardless of how complicated my chart is), but with the `y` position of labels on the `xAxis` (or `xAxes`) it seems a little broken somehow.

**Comment by pawelfus** [comment]
Hi @drmrbrewer 
I have tried some different solutions and I must stick to the first idea (improved a little). 

> Hit 'Run' several times... this now changes the y for the labels randomly between 0 and undefined, and this does change the layout significantly at the bottom (see gap between labels and chart edge):

Fixed, my changes don't affect the bottom xAxis anymore

> This (i.e. whether y of bottom labels is 0 or undefined) also seems to have a small effect on the layout at the top (see highlighted gap below, between labels and title):

I can't change it. After digging further, I consider this is a bug with `undefined`, which miscalculates available space by ~3px. However, "fixing" it generates ~1k failing tests because since ~2015 we assumed this ~3px offset is correct. 

Current solution: https://jsfiddle.net/BlackLabel/h6twmcr1/11/

**Comment by drmrbrewer** [comment]
@pawelfus the y positioning of the bottom axis label is a bit strange now isn't it?

See this example:  https://jsfiddle.net/yjhq836o/

It's the same as your recent example, but using the same `labels` object for top and bottom (not just top).  Look at the y position of the bottom labels now for column 1 vs 2 vs 3.

**Comment by pawelfus** [comment]
> @pawelfus the y positioning of the bottom axis label is a bit strange now isn't it?

I see no difference between [the current version](https://jsfiddle.net/BlackLabel/01gbza7t/) and [my PR](https://jsfiddle.net/yjhq836o/). My PR should change rendering top xAxis labels only, all other axes should remain unchanged

**Comment by pawelfus** [comment]
I might be missing something, but in the top labels, the problem was with the unpredictable height of the labels, right? As a result of this issue, using `labels.y` was problematic with multiline texts. 

As for the bottom xAxis. default `labels.y` adapts to the font size of the label (see [docs](https://api.highcharts.com/highcharts/xAxis.labels.y)).

**Comment by drmrbrewer** [comment]
@pawelfus maybe I'm missing something myself, but my understanding is that the effect of `labels.y` should be the same for both top and bottom xAxis... after all they are both the same type of axis, with one being `opposite`.

But for the bottom xAxis this is what we see for `labels.y = 0` (middle column):

![image](https://user-images.githubusercontent.com/9009792/170572708-bac03b51-b847-46c0-b801-331c43ec8435.png)
![image](https://user-images.githubusercontent.com/9009792/170572781-f873c576-ad12-485e-83fd-73b6bfb8c373.png)
![image](https://user-images.githubusercontent.com/9009792/170572817-749ffc33-2fe7-45e0-a033-143b85455b3a.png)

... and this is what we see for `labels.y = undefined` (right-most column):

![image](https://user-images.githubusercontent.com/9009792/170572934-714721b6-34ca-45b3-957f-fcaebcb53faa.png)
![image](https://user-images.githubusercontent.com/9009792/170573006-c70b2e8e-d0de-4665-ac10-84015c6c01da.png)
![image](https://user-images.githubusercontent.com/9009792/170573058-d75bec52-6d10-435d-bbdc-533e9d24b13f.png)

So for the *bottom* xAxis, the `y` positioning of the labels is different for `labels.y = 0` (middle column) and `labels.y = undefined` (right-most column).

But for the *top* xAxis, the `y` positioning of the labels is the same for `labels.y = 0` (middle column) and `labels.y = undefined` (right-most column):

![image](https://user-images.githubusercontent.com/9009792/170574237-fb563a9f-ee12-4694-aa75-456d9ad90e94.png)
![image](https://user-images.githubusercontent.com/9009792/170574273-41aa562e-449b-4b6c-893b-6f6e6f2dc745.png)
![image](https://user-images.githubusercontent.com/9009792/170574316-1a3b0b71-186b-4500-a5c3-92188a2a6a91.png)


So why does `labels.y` behave differently for top and bottom?

**Comment by drmrbrewer** [comment]
@pawelfus it's looking pretty good so far!  There's still a problem I'm struggling with though, and it may or may not be related to this particular issue, but in any case I wonder whether you had some insight:

https://jsfiddle.net/mt7fbkpy/

In the above example, the third argument to your `render()` function now controls the `font-size` used for the `xAxis` labels (not the `labels.y` value as before).  

If you look across each row (font size increasing) you can see that the gap between the labels and the xAxis is *increasing*, and the gap between the labels and the title is *decreasing*:

![image](https://user-images.githubusercontent.com/9009792/171128752-dc6679b9-012b-49a6-8cf6-e35bda4cab14.png)
![image](https://user-images.githubusercontent.com/9009792/171128819-6b5d2b83-c5f1-4fa4-9181-8e2641943904.png)
![image](https://user-images.githubusercontent.com/9009792/171128891-149f01a4-6bb6-4fac-9b26-d47c00cff5e7.png)

Ideally I'd like the axis labels to be positioned with vertical gaps above/below that are constant, or at least consistent (i.e. both increasing/decreasing with font size at the same rate)... in my use case the font size is variable and therefore it needs to look right regardless of font size.

**Comment by drmrbrewer** [comment]
@pawelfus just wondering if you have any thoughts on my last couple of comments?

**Comment by drmrbrewer** [comment]
@raf18seb is there any way to change things so that the y offset *does* take the font size into consideration, even without `useHTML: true`?  Setting `useHTML: true` breaks something else in my code... it's possible I can work around it but I thought I'd just check if there was any way that the core highcharts could position the labels more accurately in this situation even without `useHTML: true`?

**Comment by raf18seb** [comment]
@drmrbrewer, unfortunately, you're setting the different CSS `font-size` styles on your `span` elements in the `xAxis.labels.format` so, to make it work properly, your labels should be flagged as HTML elements.

I'd also go for the `getBBox()` as suggested on the Forum.
